Be smarter about mtimes when stat()'ing symlinks.

agrieve opened the pull request:
Before this change, symlinks were always followed when retrieving
mtimes. This can cause ninja to think targets are stale when outputs are
symlinks.

This patch introduces the following stat() logic for symlinks:
1. When Stat()ing inputs, use the newer of the symlink or what it points to.
- By looking at both, it maintains the pre-existing behaviour, where
  you could use a symlink as an input and ninja would know to rebuild
  when the thing it points to changes.
  1. When Stat()ing outputs, always use the timestamp of the symlink.
- If it's listed as an output, then it contains the only mtime ninja
  should care about.

Fixes Bug #1186

agrieve commented:
Note: I confirmed this fixes the Chromium Cronet Mips bot's up-to-date check by scp'ing a copy of ninja with this patch:

```
chrome-bot@slave114-c1:(Linux 14.04):~$ /b/build/scripts/slave/.recipe_deps/depot_tools/ninja -w dupbuild=err -C /b/build/slave/Android_Cronet_MIPS_Builder/build/src/out/Release cronet_package cronet_perf_test_apk cronet_sample_test_apk cronet_test_instrumentation_apk cronet_unittests cronet_unittests_apk net_unittests net_unittests_apk -j50 -d explain -n
ninja: Entering directory `/b/build/slave/Android_Cronet_MIPS_Builder/build/src/out/Release'
ninja explain: restat of output md5sum_bin_host older than most recent input obj/tools/android/md5sum/md5sum_bin_host.inputdeps.stamp (0 vs 1474466541)
ninja explain: md5sum_bin_host is dirty
ninja explain: obj/tools/android/md5sum/md5sum_bin_host.stamp is dirty
ninja explain: restat of output host_forwarder older than most recent input obj/tools/android/forwarder2/host_forwarder.inputdeps.stamp (0 vs 1474466541)
ninja explain: host_forwarder is dirty
[5/5] STAMP obj/tools/android/md5sum/md5sum.stamp
chrome-bot@slave114-c1:(Linux 14.04):~$ /b/build/scripts/slave/.recipe_deps/depot_tools/ninja -w dupbuild=err -C /b/build/slave/Android_Cronet_MIPS_Builder/build/src/out/Release cronet_package cronet_perf_test_apk cronet_sample_test_apk cronet_test_instrumentation_apk cronet_unittests cronet_unittests_apk net_unittests net_unittests_apk -j50 -d explain -n
ninja: Entering directory `/b/build/slave/Android_Cronet_MIPS_Builder/build/src/out/Release'
ninja explain: restat of output md5sum_bin_host older than most recent input obj/tools/android/md5sum/md5sum_bin_host.inputdeps.stamp (0 vs 1474466541)
ninja explain: md5sum_bin_host is dirty
ninja explain: obj/tools/android/md5sum/md5sum_bin_host.stamp is dirty
ninja explain: restat of output host_forwarder older than most recent input obj/tools/android/forwarder2/host_forwarder.inputdeps.stamp (0 vs 1474466541)
ninja explain: host_forwarder is dirty
[5/5] STAMP obj/tools/android/md5sum/md5sum.stamp
chrome-bot@slave114-c1:(Linux 14.04):~$ ./ninja -w dupbuild=err -C /b/build/slave/Android_Cronet_MIPS_Builder/build/src/out/Release cronet_package cronet_perf_test_apk cronet_sample_test_apk cronet_test_instrumentation_apk cronet_unittests cronet_unittests_apk net_unittests net_unittests_apk -j50 -d explain -n
ninja: Entering directory `/b/build/slave/Android_Cronet_MIPS_Builder/build/src/out/Release'
ninja: no work to do.
```

maximuska commented:
I haven't looked at the patch.

But as for the described logic, IMHO it would be simpler to allow
explicitly specifying in the manifest whether stat or lstat should be used
for each specific target.

How are you proposing handling broken symlinks?

On Sep 21, 2016 5:51 PM, "agrieve" <notifications@github.com> wrote:

> Before this change, symlinks were always followed when retrieving
> mtimes. This can cause ninja to think targets are stale when outputs are
> symlinks.
>
> This patch introduces the following stat() logic for symlinks:
> 1. When Stat()ing inputs, use the newer of the symlink or what it points
> to.
>
>    - By looking at both, it maintains the pre-existing behaviour, where
>    you could use a symlink as an input and ninja would know to rebuild when
>    the thing it points to changes.
>       1. When Stat()ing outputs, always use the timestamp of the symlink.
>    - If it's listed as an output, then it contains the only mtime ninja
>    should care about.
>
> Fixes Bug #1186 <https://github.com/ninja-build/ninja/issues/1186>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/ninja-build/ninja/pull/1188
> Commit Summary
>
>    - Be smarter about mtimes when stat()'ing symlinks.
>
> File Changes
>
>    - *M* src/build.cc
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-0> (15)
>    - *M* src/build_test.cc
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-1> (8)
>    - *M* src/clean.cc
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-2> (3)
>    - *M* src/clean_test.cc
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-3> (72)
>    - *M* src/disk_interface.cc
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-4> (27)
>    - *M* src/disk_interface.h
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-5> (14)
>    - *M* src/disk_interface_test.cc
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-6> (148)
>    - *M* src/graph.cc
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-7> (13)
>    - *M* src/graph.h
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-8> (12)
>    - *M* src/manifest_parser_perftest.cc
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-9> (3)
>    - *M* src/ninja.cc
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-10> (6)
>    - *M* src/test.cc
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-11> (5)
>    - *M* src/test.h
>    <https://github.com/ninja-build/ninja/pull/1188/files#diff-12> (4)
>
> Patch Links:
>
>    - https://github.com/ninja-build/ninja/pull/1188.patch
>    - https://github.com/ninja-build/ninja/pull/1188.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ninja-build/ninja/pull/1188>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFq-nI51Fvf0M921S8-Aa8-159uoagxks5qsURTgaJpZM4KC5e->
> .
>

agrieve commented:
I've added a unit test for broken links, but basically I'm proposing that we use the symlink's timestamp for this case.

If there's a case where explicitly specifying the whether to follow the symlink when checking mtimes is useful, then I'd be fine going that route as well. I don't think it's necessarily simpler though to add another thing the .ninja author needs to worry about.

I think what I've proposed here is a better default behavior at least. Specifically - it never makes sense to follow an output symlink when retrieve mtimes.

nico commented:
Hm, that's a pretty large patch for a fairly edge-case-y behavior that I wouldn't be surprised someone relied on in some surprising way. Can you make your build not use symlinks instead?

agrieve commented:
I could switch the Android build to generate wrapper scripts rather than symlinks, but in the spirit of fixing the root cause of things, I arrived here (plus I also discovered iOS creates symlinks as well). The patch is actually quite small if you ignore updating every call site of `Stat()`.

Here's why I believe this is a clear bug:

Image you have a ninja rule with 2 inputs and an output symlink to input1. If input2 has a timestamp newer than input1, ninja will forever think that the output is stale, because it's reporting input1's timestamp rather than the output symlink's timestamp.

agrieve commented:
Thought of an even simpler example:

* A rule with only one input: makesymlink.py
* An output symlink that points to a checked-in file (not listed as an input)

If the checked-in file the symlink points to is not newer than the checked-in makesymlink.py, ninja will always think the target is dirty.

agrieve commented:
🐼?

colincross commented:
The android build needs symlinks in lots of places, and we aren't going to be able to avoid that.  Right now we create them as side-effects of installing the the file they link to, but that means that ninja won't recreate them if you delete one.  Using lstat on output files would be perfect for us.

agrieve commented:
🐹

agrieve commented:
@nico - Any further thoughts here? The main change here is ~30 lines, which I don't think is that big.

rsesek commented:
I also hit this issue in Mac framework packaging and have had to work around it (https://codereview.chromium.org/2487763002/).

agrieve commented:
🎁

agrieve commented:
Another month, another ping.

agrieve commented:
@nico What could I do to convince you to look at this? Are there other owners for ninja that could take a look?

agrieve commented:
Q: Why not switching everything to lstat()?
A: Consider an existing setup like:
1. Shared network drive with BigFile.dat on it
2. Local build checkout that expects BigFile.dat to be within the checkout
3. User symlinks BigFile.dat to point to the network copy
4. Network copy of BigFile.dat changes, you'd want ninja to consider it dirty.

Q: Okay, so why not leave it as stat()?
A: Consider this setup:
1. A target depends on foo.dat, which links to /a/foo.dat
2. User changes symlink from /a/foo.dat to /b/foo.dat
3. Both /a/foo.dat and /b/foo.dat are older than the output, but because the symlink's mtime is now newer, ninja should rebuild.

Q: Why not use stamp files for output edges, then have symlinks ninja doesn't know about:
A: If someone deletes the actual symlink and not the stamp file, ninja will not know to rebuild it. Having ninja just do the right thing would be much better than having unspecified outputs.

Q: Why not make this patch smaller using a default value for Stat()'s new parameter:
A: I could make the patch a bunch smaller by making DiskInterface::kDontFollow be the default value for Stat(), but I thought making it explicit everywhere would lead to fewer bugs.

Q: Why is this issue important?
A: Because correct builds are important.

agrieve commented:
Somehow I don't see it in the history here, but the relavant Android Chrome bug for this is:
https://bugs.chromium.org/p/chromium/issues/detail?id=646165

agrieve commented:
ping

agrieve commented:
ping

nico commented on src/disk_interface.h:
```diff
@@ -44,9 +44,18 @@ struct FileReader {
 struct DiskInterface: public FileReader {
+  /// How to treat symlinks when calling Stat().
+  enum SymlinkTreatment {
+    kFollow,      // Used when the only the file contents relevant.
```
the only the

nico commented on src/disk_interface.h:
```diff
@@ -44,9 +44,18 @@ struct FileReader {
+  enum SymlinkTreatment {
+    kFollow,      // Used when the only the file contents relevant.
+    kDontFollow,  // Used for outputs.
+    kTakeNewest,  // Used for inputs.
```
Thanks for the comments here, they're very helpful.

However, aren't most nodes both outputs and inputs, yet stat()'d only once? What happens in that case?

nico commented:
Alright, let's get this in. Sorry about the very long wait. Please see the one question below.

agrieve commented:
Thanks for taking the time on this!

Updated the comments to better explain things. Rebased, and also removed kFollow. I don't think there's ever a situation where kTakeNewest wouldn't be safer to use.

nico commented:
I tried reading some docs and from what I can tell posix says "For symbolic links, the st_mode member shall contain meaningful information when used with the file type macros, and the st_size member shall contain the length of the pathname contained in the symbolic link. File mode bits and the contents of the remaining members of the stat structure are unspecified." – i.e. it's up to chance what's in the timestamp for symbolic links. That maybe feels a bit brittle?

agrieve commented:
When I was looking at this before, that had me worried as well. I came across this though:
https://linux.die.net/man/2/lstat

According to POSIX.1-2001, lstat() on a symbolic link need return valid information only in the st_size field and the file-type component of the st_mode field of the stat structure. POSIX.-2008 tightens the specification, requiring lstat() to return valid information in all fields except the permission bits in st_mode.

I'd guess by now that assuming posix-2008 compliance is okay, but really have no experience with making these sorts of judgements. wdyt?

nico commented on src/graph.cc:
```diff
@@ -27,8 +27,12 @@
+bool Node::Stat(DiskInterface* disk_interface,
+                DiskInterface::SymlinkTreatment symlink_treatment,
+                string* err) {
+  METRIC_RECORD("node stat");
```
Is this a mismerge? danalbert moved this METRIC_RECORD into disk_interface recently.

agrieve replied:
blarg, looks like it. Thanks for catching!

nico commented:
Ah thanks, I think that's fine then.

One tiny comment, but let's merge this and see how it goes. Is this the only thing needed to make the chrome/android build have clean rebuilds, or is there more work needed that's blocked on this?

Everyone else, I'm interested in feedback on this – so please let me know how your builds look with this merged (once it is merged).

agrieve commented:
Fixed up the mismerge. This is the only thing required to enable the clean rebuild check on the android bots (unless a new source of uncleanliness has crept in in the meantime)

atetubou commented on src/disk_interface.cc:
```diff
@@ -181,7 +184,22 @@ TimeStamp RealDiskInterface::Stat(const string& path, string* err) const {
+    result = lstat(path.c_str(), &st);
+    break;
+  case kTakeNewest:
+  default:
```
need  this default label?

agrieve replied:
Lint likes it to be there.

agrieve commented:
Ready for action?

nico commented:
agrieve and I discussed this change in person some earlier this week and mostly convinced ourselves that it'd break some existing build files. We then tried to understand what exactly goes wrong without this patch and failed to recreate that too. So for now, it's not clear if anything needs to be done in ninja here.

dpranke commented:
I wonder if this issue was partially a result of crbug.com/748623, where the way Chromium managed testing patches on the bots confused the mtimes of files, and running `git update-index --refresh` has maybe made the problem go away?

nico commented:
I found a repro!

Let's say you're trying to build an executable, and you also want a symlink to that executable under a different name (think e.g. clang and clang-cl). In gn, you might express this as a target that creates the symlink, and then make the actual executable link depend on the symlink target (the symlink's dest does after all not have to exist when the symlink's created). gn will then make a build rule that creates a symlink, then a stamp file that depends on the symlink for the symlink action, and then a link rule for the executable, order-only depending on the stamp, like so:

```
rule symlink
  command = ln -sf lld lld-link
  restat = 1
build lld-link: symlink

rule stamp
  command = touch $out
  restat = 1
build lld-link.stamp: stamp lld-link

rule link
  command = sleep 2 && touch $out
  restat = 1
build lld: link || lld-link.stamp
```

With that setup:

```
Nicos-MacBook-Pro:foo thakis$ ninja -d explain | cat 
ninja explain: output lld-link doesn't exist
ninja explain: lld-link is dirty
ninja explain: output lld doesn't exist
ninja explain: output lld-link.stamp doesn't exist
[1/3] ln -sf lld lld-link
[2/3] touch lld-link.stamp
[3/3] sleep 2 && touch lld
Nicos-MacBook-Pro:foo thakis$ ninja -d explain | cat 
ninja explain: restat of output lld-link.stamp older than most recent input lld-link (1522896947 vs 1522896949)
[1/1] touch lld-link.stamp
Nicos-MacBook-Pro:foo thakis$ ninja -d explain | cat 
ninja: no work to do.
```

nico commented:
Since we're closing pull requests we're not planning on merging nowadays, let's close this out given the discussion.
